### PR TITLE
Converted @typechecked to recompile the target function with type checks

### PR DESCRIPTION
**Pull request by agronholm**
We no longer wrap the callable, thus eliminating the extra stack frame.

### CODE REVIEW

File: src/typeguard/_decorators.py [hunk 1]
```diff
@@ -31,6 +27,60 @@ def typeguard_ignore(f: _F) -> _F:
 T_CallableOrType = TypeVar("T_CallableOrType", bound=Callable[..., Any])
 
 
+def make_cell():
+    value = None
+    return (lambda: value).__closure__[0]
+
+
+def instrument(f: T_CallableOrType) -> Callable | str:
+    if not getattr(f, "__annotations__", None):
```
Comment (marcelroed) [inline src/typeguard/_decorators.py:36]: Why not use `hasattr`?

File: src/typeguard/_decorators.py [hunk 2]
```diff
@@ -31,6 +27,60 @@ def typeguard_ignore(f: _F) -> _F:
...
+    module_code = compile(module_ast, module.__file__, "exec", dont_inherit=True)
+    new_code = module_code
+    for level, name in enumerate(target_path):
+        for const in new_code.co_consts:
+            if isinstance(const, CodeType):
+                if const.co_name == name:
+                    new_code = const
+                    break
+        else:
+            return "cannot find the target function in the AST"
```
Comment (marcelroed) [inline src/typeguard/_decorators.py:58]: Could create an error type and raise it here

File: src/typeguard/_decorators.py (hunk 1 above)
Comment (marcelroed) [reply src/typeguard/_decorators.py:36]: I suppose this also catches the case where `__annotations__ is None`?

Comment (agronholm) [reply src/typeguard/_decorators.py:36]: > I suppose this also catches the case where `__annotations__ is None`?

Indeed.

**Comment by agronholm** [comment]
Quick reminder: [the other PR](https://github.com/agronholm/typeguard/pull/294) has all this plus the checking of annotated variable assignments.

**Comment by agronholm** [comment]
Found a bug: in the case of `@overload`, it instruments one of the overloads and not the implementation.

**Comment by agronholm** [comment]
Fixed that in the other PR.

**Comment by agronholm** [comment]
@marcelroed I'm going to have to move on with these changes, so I can't wait much longer for your feedback. Let me know if you need more time.

**Comment by marcelroed** [comment]
Go for it! I looked through it and it looks great!

**Comment by agronholm** [comment]
Great, thanks!

**Comment by agronholm** [comment]
> Go for it! I looked through it and it looks great!

Does this mean you also looked at variable assignment checking?

**Comment by marcelroed** [comment]
Yep, I added some comments now, but I couldn't find any issues without going very deep. To be honest, any `ast` code is a bit opaque without stepping through with a debugger.

**Pull request merged**